### PR TITLE
refactor: use pointer types more in machine config structs

### PR DESCRIPTION
**Pull request by smira**
There should be no functional change with this PR.

The primary driver is supporting stragetic merge configuration patches.
For such type of patches machine config should be loaded from incomplete
fragments, so it becomes critically important to distinguish between a
field having zero value vs. field being set in YAML.

E.g. with following struct:

```go
struct { AEnabled *bool `yaml:"a"` }
```

It's possible to distinguish between:

```yaml
a: false
```

and no metion of `a` in YAML.

Merging process trewats zero values as "not set" (skips them when
merging), so it's important to allow overriding value to explicit
`false`.

Signed-off-by: Andrey Smirnov <andrey.smirnov@talos-systems.com>

---
This change is [Reviewable](https://reviewable.io/reviews/siderolabs/talos/5842)

### CODE REVIEW

**Review by frezbo — APPROVED** [review]

File: pkg/machinery/config/types/v1alpha1/v1alpha1_apiserverconfig.go [hunk 1]
```diff
@@ -40,7 +40,7 @@ func (a *APIServerConfig) Env() Env {
 
 // DisablePodSecurityPolicy implements the config.APIServer interface.
 func (a *APIServerConfig) DisablePodSecurityPolicy() bool {
-	return a.DisablePodSecurityPolicyConfig
+	return a.DisablePodSecurityPolicyConfig == nil || *a.DisablePodSecurityPolicyConfig
```
Comment (smira) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_apiserverconfig.go:43]: wrong!

File: pkg/machinery/config/types/v1alpha1/v1alpha1_discoveryconfig.go [hunk 2]
```diff
@@ -10,12 +10,12 @@ import (
 )
 
 // Enabled implements the config.ClusterDiscovery interface.
-func (c ClusterDiscoveryConfig) Enabled() bool {
-	return c.DiscoveryEnabled
+func (c *ClusterDiscoveryConfig) Enabled() bool {
+	return c.DiscoveryEnabled != nil && *c.DiscoveryEnabled
```
Comment (Unix4ever) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_discoveryconfig.go:14]: Would suggest using a helper function like:

```
func IsTrue(value *bool) bool {
  return value != nil && *value
}
```

Comment (smira) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_discoveryconfig.go:14]: we could even do a generic function for `*T`... I will try that... or we even had one

Comment (smira) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_discoveryconfig.go:14]: https://github.com/siderolabs/go-pointer/blob/master/pointer.go#L13

Comment (smira) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_discoveryconfig.go]: updated

**Review by Unix4ever — APPROVED** [review]
:cool:

**Pull request merged**